libnetwork/service_linux.go

```diff
@@ -359,13 +360,21 @@ func programIngress(gwIP net.IP, ingressPorts []*PortConfig, isDelete bool) erro
 			}
 		}
 
-		if !iptable.Exists(iptables.Filter, "FORWARD", "-j", ingressChain) {
-			if err := iptable.RawCombinedOutput("-I", "FORWARD", "-j", ingressChain); err != nil {
-				return fmt.Errorf("failed to add jump rule to %s in filter table forward chain: %v", ingressChain, err)
+		// The DOCKER-FORWARD chain is created by the bridge driver on startup. It's a stable place to
+		// put the jump to DOCKER-INGRESS (nothing else will ever be inserted before it, and the jump
+		// will precede the bridge driver's other rules).
+		if !iptable.Exists(iptables.Filter, bridge.DockerForwardChain, "-j", ingressChain) {
+			if err := iptable.RawCombinedOutput("-I", bridge.DockerForwardChain, "-j", ingressChain); err != nil {
+				return fmt.Errorf("failed to add jump rule to %s in filter table %s chain: %v",
+					ingressChain, bridge.DockerForwardChain, err)
 			}
-			// The jump to DOCKER-USER needs to be before the jump to DOCKER-INGRESS.
-			if err := setupUserChain(iptables.IPv4); err != nil {
-				log.G(context.TODO()).Warnf("Failed to restore "+userChain+" after creating "+ingressChain+": %v", err)
+		}
+		// Remove the jump from FORWARD to DOCKER-INGRESS, if it was created there by a version of
+		// the daemon older than 28.0.1.
+		// FIXME(robmry) - should only do this once, on startup.
+		if iptable.Exists(iptables.Filter, "FORWARD", "-j", ingressChain) {
```

Comment:
VERY unlikely to be relevant in this spot and _no need to update_ 👈 , but Go (GoDoc) has some extra semantics for "special" comments (most notably `Deprecated:` but ISTR also to some extent `TODO` / `FIXME` comments); to prevent such comments from being considered part of the comment itself, they expect there to be an empty (comment-)line before them, and for them to be last in the comment block, e.g.;

```suggestion
		// the daemon older than 28.0.1.
		//
		// FIXME(robmry) - should only do this once, on startup.
		if iptable.Exists(iptables.Filter, "FORWARD", "-j", ingressChain) {
```

In this case it unlikely matters as it's a comment internal to the function (which Go itself doesn't really bother with), but I try to stick to that convention when writing these, also to make sure IDEs can spot them (which tend to try to follow Go conventions, although usually are slightly more permissive).

Reply:
Oh, thank you ... will do for next time.

I should have found somewhere better to put that delete, but this seemed safest right now - it mustn't happen on a live-restore (if swarm supports that), and I didn't want to miss any other special case. It can be tidied up as we shift over to nftables.